src/main/java/com/meta/chatbridge/llm/HuggingFaceLlamaPlugin.java

```diff
@@ -0,0 +1,187 @@
+            String role = m.get("role").textValue();
+            if (role.equals("system")) {
+                continue; // system has already been counted
+            }
+            totalTokens += tokenCount(m);
+            if (totalTokens > config.maxInputTokens()) {
+                break;
+            }
+            output.insert(0, m);
+        }
+        if (hasSystemMessage) {
+            output.insert(0, systemMessage);
+        }
+
+        if ((hasSystemMessage && output.size() <= 1) || output.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return Optional.of(output);
+    }
+
+    @Override
+    public T handle(MessageStack<T> messageStack) throws IOException {
+        T fromUser = messageStack.tail();
+
+        ObjectNode body = MAPPER.createObjectNode();
+        ObjectNode params = MAPPER.createObjectNode();
+
+        config.topP().ifPresent(v -> params.put("top_p", v));
+        config.temperature().ifPresent(v -> params.put("temperature", v));
+        config.maxOutputTokens().ifPresent(v -> params.put("max_new_tokens", v));
+
+        body.set("parameters", params);
+
+        String prompt = getPrompt(messageStack);
+
+        body.put("inputs", prompt);
+
+        String bodyString;
+        try {
+            bodyString = MAPPER.writeValueAsString(body);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e); // this should be impossible
+        }
+        Response response =
+                Request.post(endpoint)
+                        .bodyString(bodyString, ContentType.APPLICATION_JSON)
+                        .setHeader("Authorization", "Bearer " + config.apiKey())
+                        .execute();
+
+        JsonNode responseBody = MAPPER.readTree(response.returnContent().asBytes());
+        String allGeneratedText = responseBody.get(0).get("generated_text").textValue();
+        String llmResponse = allGeneratedText.trim().replace(prompt.trim(), "");
+        Instant timestamp = Instant.now();
+
+        return messageStack.newMessageFromBot(timestamp, llmResponse);
+    }
+
+    private String getPrompt(MessageStack<T> MessageStack) {
+        List<String> texts = new ArrayList<>();
+        if(config.systemMessage().isPresent()){
+            texts.add("<s>[INST] <<SYS>>\n" + config.systemMessage() + "\n<</SYS>>\n\n");
+        } else {
+            texts.add("<s>[INST] ");
+        }
+
+        // The first user input is _not_ stripped
+        boolean doStrip = false;
+        Message.Role lastMessageSender = Message.Role.SYSTEM;
+
+        for (T message : MessageStack.messages()) {
+            String text = doStrip ?  message.message().trim() : message.message();
+            Message.Role user = message.role();
+            boolean isUser = user.equals(Message.Role.USER);
+            if(isUser){
+                doStrip = true;
+            }
+
+            if(isUser && lastMessageSender.equals(Message.Role.ASSISTANT)){
```

Comment:
nit: you can compare enums with `==`
```suggestion
            if(isUser && lastMessageSender == Message.Role.ASSISTANT){
```